### PR TITLE
[Flax] Add other BERT classes

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds the other BERT model classes for Flax.

Also the following checkpoints have been uploaded for Flax:
- https://huggingface.co/bert-base-cased
- https://huggingface.co/bert-large-cased
- https://huggingface.co/bert-base-uncased
- https://huggingface.co/bert-large-uncased
- https://huggingface.co/bert-large-uncased-whole-word-masking-finetuned-squad

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 1]
```diff
@@ -597,8 +695,8 @@ def setup(self):
         self.bert = FlaxBertModule(
             config=self.config,
             add_pooling_layer=False,
+            dtype=self.dtype,
         )
-        self.dropout = nn.Dropout(rate=self.config.hidden_dropout_prob)
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_flax_bert.py:601]: Nice catch!

**Review by LysandreJik — APPROVED** [review]
Cool, LGTM! Great work implementing all of these. Will RoBERTa follow?

It is so similar to the PyTorch implementation it seems a script could take care of the implementation by copying the PyTorch one and replacing a few strings!

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 2]
```diff
@@ -551,6 +575,79 @@ def __call__(self, input_ids, attention_mask, token_type_ids, position_ids, dete
...
+
+    def setup(self):
+        self.bert = FlaxBertModule(
+            config=self.config,
+            dtype=self.dtype,
+        )
+        self.cls = FlaxBertPreTrainingHeads(
+            config=self.config,
+            dtype=self.dtype,
+        )
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:636]:
```suggestion
        self.bert = FlaxBertModule(config=self.config, dtype=self.dtype)
        self.cls = FlaxBertPreTrainingHeads(config=self.config, dtype=self.dtype)
```
How come this takes so much space? It all fits in one line.

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 3]
```diff
@@ -611,7 +709,365 @@ def __call__(
...
+
+class FlaxBertForNextSentencePredictionModule(nn.Module):
+    config: BertConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.bert = FlaxBertModule(
+            config=self.config,
+            dtype=self.dtype,
+        )
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:767]:
```suggestion
        self.bert = FlaxBertModule(config=self.config, dtype=self.dtype)
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 4]
```diff
@@ -611,7 +709,365 @@ def __call__(
...
+
+class FlaxBertForSequenceClassificationModule(nn.Module):
+    config: BertConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.bert = FlaxBertModule(
+            config=self.config,
+            dtype=self.dtype,
+        )
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:835]:
```suggestion
        self.bert = FlaxBertModule(config=self.config, dtype=self.dtype)
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 5]
```diff
@@ -611,7 +709,365 @@ def __call__(
...
+        self.dropout = nn.Dropout(rate=self.config.hidden_dropout_prob)
+        self.classifier = nn.Dense(
+            self.config.num_labels,
+            dtype=self.dtype,
+        )
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:840]:
```suggestion
        self.classifier = nn.Dense(self.config.num_labels, dtype=self.dtype)
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 6]
```diff
@@ -611,7 +709,365 @@ def __call__(
...
+
+class FlaxBertForMultipleChoiceModule(nn.Module):
+    config: BertConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.bert = FlaxBertModule(
+            config=self.config,
+            dtype=self.dtype,
+        )
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:909]:
```suggestion
        self.bert = FlaxBertModule(config=self.config, dtype=self.dtype)
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 7]
```diff
@@ -611,7 +709,365 @@ def __call__(
...
+class FlaxBertForQuestionAnsweringModule(nn.Module):
+    config: BertConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.bert = FlaxBertModule(
+            config=self.config,
+            dtype=self.dtype,
+            add_pooling_layer=False,
+        )
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:1059]:
```suggestion
        self.bert = FlaxBertModule(config=self.config, dtype=self.dtype, add_pooling_layer=False)
```

File: src/transformers/models/bert/modeling_flax_bert.py [hunk 8]
```diff
@@ -611,7 +709,365 @@ def __call__(
...
+class FlaxBertForTokenClassificationModule(nn.Module):
+    config: BertConfig
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        self.bert = FlaxBertModule(
+            config=self.config,
+            dtype=self.dtype,
+            add_pooling_layer=False,
+        )
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_flax_bert.py:989]:
```suggestion
        self.bert = FlaxBertModule(config=self.config, dtype=self.dtype, add_pooling_layer=False)
```

**Review by sgugger — APPROVED** [review]
Looks great! I don't know what happened to your formatter but we can definitely have most of submodule definitions fit on one line ;-)

**Comment by avital** [comment]
> It is so similar to the PyTorch implementation it seems a script could take care of the implementation by copying the PyTorch one and replacing a few strings!

@marcvanzee and I were also wondering about this in general -- is there a 80/20 solution that requires user input in some cases? It would have to not introduce silent errors (e.g. a model that seems to run the same but differs in some hard-to-find way).

**Comment by merrymercy** [comment]
It seems this PR broke the flax example : https://github.com/huggingface/transformers/blob/master/examples/language-modeling/run_mlm_flax.py

This line (https://github.com/huggingface/transformers/blob/896d7be97401a85dc0ffc5460afd707e8e092781/examples/language-modeling/run_mlm_flax.py#L577) will raise the error
```
TypeError: __init__() got an unexpected keyword argument 'dropout_rate'
```

In addition, this line https://github.com/huggingface/transformers/blob/896d7be97401a85dc0ffc5460afd707e8e092781/src/transformers/models/bert/modeling_flax_bert.py#L254 uses an undefined variable `self.dropout_rate`.

**Comment by merrymercy** [comment]
@patrickvonplaten It seems this PR broke the flax example: https://github.com/huggingface/transformers/blob/master/examples/language-modeling/run_mlm_flax.py

This line (https://github.com/huggingface/transformers/blob/896d7be97401a85dc0ffc5460afd707e8e092781/examples/language-modeling/run_mlm_flax.py#L577) will raise the error
```
TypeError: __init__() got an unexpected keyword argument 'dropout_rate'
```

In addition, this line https://github.com/huggingface/transformers/blob/896d7be97401a85dc0ffc5460afd707e8e092781/src/transformers/models/bert/modeling_flax_bert.py#L254 uses an undefined variable `self.dropout_rate`.

**Comment by merrymercy** [comment]
I am very interested in the jax/flax integration, could you also take a look at my PR (https://github.com/huggingface/transformers/pull/10796)?
If you are collaborative and welcome contributions from my side, I can contribute more flax examples.

**Pull request merged**